src/include/server/mir/frontend/surface_stack.h

```diff
@@ -38,6 +43,10 @@ class SurfaceStack
     virtual void add_observer(std::shared_ptr<scene::Observer> const& observer) = 0;
     virtual void remove_observer(std::weak_ptr<scene::Observer> const& observer) = 0;
 
+    /// Returns the stacking order of the given surfaces, bottom to top
+    /// If a surface is not known to the surface stack it will not be in the returned list
+    virtual auto stacking_order_of(scene::SurfaceSet const& surfaces) const -> scene::SurfaceList = 0;
```

Comment:
Performance nit: it seems that there are likely to be very few surfaces in the `SurfaceSet`, which might make the set higher-overhead than just a plain vector?

I'm not sure how this is going to be called, but maybe it should take a `scene::SurfaceList` by value and return a `scene::SurfaceList`?

Reply:
I still think a set is a better idea. Here's why:
1. I'm not sure how large a set must be before lookup is faster than linearly searching a vector. I expect the X11 fronted to regularly call this with all X11 surfaces, which could be quite a few.
2. This function should only be called once or twice by different subsystems when surfaces are re-ordered. Not many times each frame. This means if the common case is slightly less performant than optimal, it wont significantly hurt Mir.
3. If we ever *do* need to call this on a large number of surfaces, the cost of incorrectly choosing a vector could be *very* high. The cost of incorrectly choosing a set will always be low. Regardless of what data structure we use, this function will be negligibly fast on small data sizes.
4. The argument is conceptually a set, so giving it that type makes it's nature more obvious to the programmer. This may be worth giving up if the performance gains were significant, but see 1-3.

Reply:
If this *does* get called with lots of surfaces, then we can sort the vector and do a binary search for the same cost as constructing the `std::map` and with the same number of operations for `::find()` and better memory-locality. Depending on how this gets (does the caller maintain a `std::set` of surfaces, and pass that in, or is one constructed for the purposes of this call?) called we can also optimize away one of the memory allocations.

But also this is somewhat of a pedantic sidebar :wink:

Reply:
I'm of the opinion this is a language/library level problem. Wanting to collect a large or small number of items and quickly check if an item is among them is literally what `std::set` is built for. If it really is so bad at it's job that we can't use it for that purpose than we shouldn't muddy window stacking logic with workarounds. We should build a new data structure that does the same thing but doesn't suck.

Reply:
If we did that it would be practical to do performance tests and determine in what situations our solution is working. As it is, doing actual profiling for every tree set/hash set/vector/sorted vector question is impractical, so we end up debating based on hunches. While your hunches are probably better than mine due to experience, neither are hard data.

Reply:
Hmm... my data structure knowledge is somewhat incomplete, but isn't a b-tree based set what we really need? It seems it's incompatible with the STL because iterator invalidation details, but it would have the performance characteristics we want, no? Cache-friendly and as fast as a sorted vector for a small number of elements, and as fast as a binary tree set for a large number of elements.

Reply:
Like most problems, I believe "rewrite Mir in Rust" is the optimal solution

Reply:
Yup, `BTreeSet` would be a pretty good candidate. We'll only ever be dealing with many fewer than 1,000 windows, so the asymptotic behaviour isn't terribly interesting for us, but this does indeed give us a better off-the-shelf option.

The `std::set` API makes various promises that we don't care about, so can't perform optimally for us.

Oh, hai! Google searching for “C++ small set” leads one to [Boost.Containers' flat_set](https://www.boost.org/doc/libs/1_73_0/doc/html/container/non_standard_containers.html#container.non_standard_containers.flat_xxx).

We could, if we wanted, drop that in.